### ISSUE TITLE
Ability to place signals along the whole rail line with fixed distance

### ISSUE THREAD

**Issue by Razip** [OpenRCT2/OpenLoco#139]
It's very annoying and time-consuming that we have to put signals manually.

For instance, in OpenTTD, you can set the distance between the signals in the Signal Selection menu, then, after placing a signal of necessary type, you just need to hold Ctrl, then drag the mouse in necessary direction and voila, the signal have been placed along the rest of the line with the same distance between each two.

**Comment by auge8472** [comment]
Another UI element for the signals is the toggling of the direction of a signal by only clicking into the tile. No pixel accurate working with the mouse pointer necessary.

**Comment by LeftofZen** [comment]
The lack of this feature has been annoying me a lot when playing the game (combined with no chain signals), so I'll have a crack at implementing it, pending all the relevant code parts are decompiled.

**Comment by LeftofZen** [comment]
Looks like the `createSignal` command hasn't been decompiled yet :(
![image](https://user-images.githubusercontent.com/7483209/146850839-028c2b9b-ec2e-463a-a564-e784d7881917.png)

**Comment by LeftofZen** [comment]
Looks like the `createSignal` game command hasn't been decompiled yet :(
![image](https://user-images.githubusercontent.com/7483209/146850913-4eb85617-1154-4784-b699-b375eeccc02a.png)

**Comment by LeftofZen** [comment]
I tried repeating the signal placement command in `SignalTab::onToolDown()` after the initial signal placement, but the game crashes in Interop.cpp when the subsequent `GameCommand::DoCommand()` calls happen. I'm fairly sure I've got all the args right so this is beyond me at this stage.

**Comment by LeftofZen** [comment]
I realised the `All connected track` options for the powered 3rd rail/overhead wires are really similar to what we want here and we'd be able to reuse that code, but that command, `createTrackMod`, also hasn't been decompiled yet.

**Comment by duncanspumpkin** [comment]
@LeftofZen lets see your code you don't need to have these functions implemented to be able to use them. You are probably just making some simple mistake.

**Comment by LeftofZen** [comment]
@duncanspumpkin Yeah I know I don't have to decompile functions to call them, I was still brainstorming when I wrote that first comment about `createSignal` and it's wrong in hindsight. However since I was unable to figure out how to get the adjacent track connections, which was what my 2nd comment referenced, I then realised I could look at the code in `createTrackMod` since that does exactly what we want, but that command isn't decompiled and that's the one I would need to copy to use for the signal placement.

Anyways I just pushed my experiments to a branch if you do wanna take a look, the part I'm uncertain about was how to use `getTrackConnections()` and the surrounding code I wrote/copied in SignalTab.cpp around line 280.

https://github.com/LeftofZen/OpenLoco/tree/%23139-auto-place-rail-signals

**Comment by duncanspumpkin** [comment]
I think you basically have it but you are resetting the location each loop when you shouldn't be. So it has the correct location for the first loop of it then it goes wrong. It's also walking backwards along the track instead of forwards. Unsure if that is an issue.

**Comment by LeftofZen** [comment]
Thanks for the tips, I believe I figured out the location problem, but I couldn't figure out how to reverse the direction/make it follow the direction of a 1-way signal. I'll probably work on something easier for me and come back to this later.

**Comment by Martinocom** [comment]
This would be REALLY useful, it's one of the most annoying things to do after creating a track.
